predictors.go

```diff
@@ -0,0 +1,240 @@
+package kong
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/posener/complete"
+)
+
+// Predictor implements a predict method, in which given command line arguments returns a list of options it predicts.
+// Taken from github.com/posener/complete
+type Predictor interface {
+	Predict(PredictorArgs) []string
+}
+
+// PredictorArgs describes command line arguments used by a Predictor to predict options.
+// Taken from Args in github.com/posener/complete
+type PredictorArgs struct {
+	// All lists of all arguments in command line (not including the command itself)
+	All []string
+	// Completed lists of all completed arguments in command line,
+	// If the last one is still being typed - no space after it,
+	// it won't appear in this list of arguments.
+	Completed []string
+	// Last argument in command line, the one being typed, if the last
+	// character in the command line is a space, this argument will be empty,
+	// otherwise this would be the last word.
+	Last string
+	// LastCompleted is the last argument that was fully typed.
+	// If the last character in the command line is space, this would be the
+	// last word, otherwise, it would be the word before that.
+	LastCompleted string
+}
+
+// NewPredictor returns a Predictor that runs the provided function.
+func NewPredictor(fn func(PredictorArgs) []string) Predictor {
+	return predictorFunc(fn)
+}
+
+type predictorFunc func(PredictorArgs) []string
+
+func (p predictorFunc) Predict(args PredictorArgs) []string {
+	if p == nil {
+		return nil
+	}
+	return p(args)
+}
+
+// PredictNothing returns a nil Predictor that indicates no prediction is to be made.
+func PredictNothing() Predictor {
+	return nil
+}
+
+// PredictAnything returns a Predictor that expects something, but nothing particular, such as a number or arbitrary name.
+func PredictAnything() Predictor {
+	return NewPredictor(func(PredictorArgs) []string {
+		return nil
+	})
+}
+
+//PredictSet returns a Predictor that predicts provided options
+func PredictSet(options ...string) Predictor {
+	return predictorFunc(func(args PredictorArgs) []string {
+		return options
+	})
+}
```

Comment:
Please add predictors for the builtin types `file` `existingfile` and `path`, as well as `enum` (if not already).

This would be a) very useful and b) serve as a good example of how to write predictors that do more than return a static set.